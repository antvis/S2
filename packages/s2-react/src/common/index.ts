export * from './constant';
export * from './i18n';
export * from './icons';
export * from './react-element';
