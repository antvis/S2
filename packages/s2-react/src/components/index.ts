export { BaseSheet } from './sheets/base-sheet';
export { TableSheet } from './sheets/table-sheet';
export { GridAnalysisSheet } from './sheets/grid-analysis-sheet';
export { StrategySheet } from './sheets/strategy-sheet';
export { AdvancedSort } from './advanced-sort';
export { DrillDown } from './drill-down';
export { TooltipComponent } from './tooltip';
export { Switcher } from './switcher';
export * from './sheets';
export * from './sheets/interface';
