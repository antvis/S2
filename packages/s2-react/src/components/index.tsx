import React from 'react';
import debounceRender from 'react-debounce-render';
import { BaseSheet } from './sheets/base-sheet';
import { TabularSheet } from './sheets/tabular-sheet';
import { TableSheet } from './sheets/table-sheet';
import { SpreadsheetProps } from './sheets/interface';
export { DrillDown } from './drill-down';
export { Switcher } from './switcher';
export { AdvancedSort } from './advanced-sort';
export * from './sheets';

export const SheetComponent: React.FC<SpreadsheetProps> = debounceRender(
  (props: SpreadsheetProps) => {
    const { sheetType } = props;
    switch (sheetType) {
      case 'table':
        return <TableSheet {...props} />;
      case 'tabular':
        return <TabularSheet {...props} />;
      default:
        return <BaseSheet {...props} />;
    }
  },
  100,
);
