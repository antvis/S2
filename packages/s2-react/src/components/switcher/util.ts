import { filter, flatten, isNil, map, mapValues } from 'lodash';
import { DraggableLocation } from 'react-beautiful-dnd';
import { getClassNameWithPrefix } from '@antv/s2';
import {
  FieldType,
  MAX_DIMENSION_COUNT,
  SWITCHER_FIELDS,
  SWITCHER_PREFIX_CLS,
} from './constant';
import {
  SwitcherItem,
  SwitcherResult,
  SwitcherState,
  SwitcherFields,
  SwitcherResultItem,
} from './interface';

export const getSwitcherClassName = (...classNames: string[]) =>
  getClassNameWithPrefix(SWITCHER_PREFIX_CLS, ...classNames);

export const getNonEmptyFieldCount = (fields: SwitcherFields) => {
  return SWITCHER_FIELDS.reduce(
    (sum, field) => sum + (isNil(fields[field]) ? 0 : 1),
    0,
  );
};

export const getMainLayoutClassName = (nonEmptyCount: number) => {
  switch (nonEmptyCount) {
    case 1:
      return getSwitcherClassName('content', 'one-dimension');
    case 2:
      return getSwitcherClassName('content', 'two-dimensions');
    default:
      return getSwitcherClassName('content', 'three-dimensions');
  }
};

export const shouldCrossRows = (nonEmptyCount: number, type: FieldType) =>
  nonEmptyCount < MAX_DIMENSION_COUNT || type === FieldType.Values;

export const moveItem = (
  source: SwitcherItem[],
  destination: SwitcherItem[],
  droppableSource: DraggableLocation,
  droppableDestination: DraggableLocation,
): SwitcherState => {
  // change order in same column
  if (droppableDestination.droppableId === droppableSource.droppableId) {
    const updatingDestination = [...destination];
    const [removed] = updatingDestination.splice(droppableSource.index, 1);
    updatingDestination.splice(droppableDestination.index, 0, removed);
    return {
      [droppableDestination.droppableId]: updatingDestination,
    };
  }
  // move to other column
  const updatingSource = [...source];
  const updatingDestination = [...destination];

  const [removed] = updatingSource.splice(droppableSource.index, 1);
  updatingDestination.splice(droppableDestination.index, 0, removed);

  return {
    [droppableSource.droppableId]: updatingSource,
    [droppableDestination.droppableId]: updatingDestination,
  };
};

export const checkItem = (
  source: SwitcherItem[],
  checked: boolean,
  id: string,
  parentId?: string,
): SwitcherItem[] => {
  const target: SwitcherItem = {
    ...source.find((item) => item.id === (parentId ?? id)),
  };

  // 有 parentId 时，说明是第二层级的改变
  if (parentId) {
    target.children = map(target.children, (item) => ({
      ...item,
      checked: item.id === id ? checked : item.checked,
    }));
  } else {
    target.checked = checked;
    target.children = map(target.children, (item) => ({
      ...item,
      checked,
    }));
  }

  return source.map((item) => (item.id === target.id ? target : item));
};

export const generateSwitchResult = (state: SwitcherState): SwitcherResult => {
  const generateFieldResult = (items: SwitcherItem[]): SwitcherResultItem => {
    const flattenValues = (list: SwitcherItem[]) =>
      flatten(
        map(list, ({ children, ...rest }) => {
          return [{ ...rest }, ...flattenValues(children)];
        }),
      );

    const allItems = flattenValues(items);

    //  get all hidden values
    const hideItems = filter(
      allItems,
      (item: SwitcherItem) => item.checked === false,
    );
    return {
      items: allItems,
      hideItems,
    };
  };

  return {
    [FieldType.Rows]: generateFieldResult(state[FieldType.Rows]),
    [FieldType.Cols]: generateFieldResult(state[FieldType.Cols]),
    [FieldType.Values]: generateFieldResult(state[FieldType.Values]),
  };
};

export const getSwitcherState = (fields: SwitcherFields): SwitcherState =>
  mapValues(fields, 'items');
