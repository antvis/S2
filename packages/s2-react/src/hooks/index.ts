export * from './usePagination';
export * from './useResize';
export * from './useLoading';
export * from './useEvents';
export * from './useSpreadSheet';
