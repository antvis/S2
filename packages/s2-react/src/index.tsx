// export * from '@antv/s2';
import React, { FC } from 'react';
import './index.less';
import './index.module.less';

export const Entry: FC = () => {
  return <div>Entry</div>;
};
