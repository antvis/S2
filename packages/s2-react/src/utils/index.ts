export * from './drill-down';
export * from './options';
