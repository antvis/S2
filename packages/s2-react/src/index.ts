// https://github.com/microsoft/TypeScript/issues/47663#issuecomment-1270716220
import type {} from 'rc-pagination';
import './utils/extendLocale';

export * from './components';
