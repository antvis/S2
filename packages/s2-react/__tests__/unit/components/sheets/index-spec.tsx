import {
  PivotSheet,
  TableSheet,
  type S2DataConfig,
  type SpreadSheet,
} from '@antv/s2';
import type { SheetType } from '@antv/s2-shared';
import { render, waitFor } from '@testing-library/react';
import React from 'react';
import { SheetComponent, type SheetComponentsProps } from '../../../../src';
import { getContainer, renderComponent } from '../../../util/helpers';

describe('<SheetComponent/> Tests', () => {
  describe('Render Tests', () => {
    let container: HTMLDivElement;

    beforeEach(() => {
      container = getContainer();
    });

    afterEach(() => {
      container.remove();
    });

    const sheetTypes: SheetType[] = [
      'pivot',
      'table',
      'strategy',
      'gridAnalysis',
      'editable',
    ];

    const commonSheetProps: SheetComponentsProps = {
      // CI 环境和 本地 DPR 不一致
      options: {
        width: 200,
        height: 200,
        transformCanvasConfig() {
          return {
            devicePixelRatio: 2,
          };
        },
      },
      dataCfg: null as unknown as S2DataConfig,
      showPagination: true,
      header: {
        switcher: { open: true },
        export: { open: true },
        advancedSort: { open: true },
      },
    };

    test.each(sheetTypes)(
      'should render successfully for %s sheet',
      async (sheetType) => {
        function init() {
          renderComponent(
            <SheetComponent sheetType={sheetType} {...commonSheetProps} />,
            container,
          );
        }

        await waitFor(() => {
          expect(init).not.toThrow();
        });
      },
    );

    test.each(sheetTypes)(
      'should render %s sheet by snapshot',
      async (sheetType) => {
        const { asFragment } = render(
          <SheetComponent sheetType={sheetType} {...commonSheetProps} />,
        );

        await waitFor(() => {
          expect(asFragment()).toMatchSnapshot();
        });
      },
    );

    test.each(sheetTypes)(
      'should render and destroy for %s sheet',
      async (sheetType) => {
        let onMountedRef: SpreadSheet;

        const onMounted = jest.fn((instance) => {
          onMountedRef = instance;
        });
        const onDestroy = jest.fn();

        const unmount = renderComponent(
          <SheetComponent
            sheetType={sheetType}
            options={{ width: 200, height: 200 }}
            dataCfg={null as unknown as S2DataConfig}
            onMounted={onMounted}
            onDestroy={onDestroy}
          />,
          container,
        );

        await waitFor(() => {
          expect(onMounted).toHaveBeenCalledWith(onMountedRef!);
          expect(onDestroy).not.toHaveBeenCalled();
        });

        unmount?.();

        await waitFor(() => {
          expect(onDestroy).toHaveBeenCalledTimes(1);
        });
      },
    );

    test('should get latest instance after sheet type changed', async () => {
      let spreadSheet: SpreadSheet;

      function Component({
        sheetType,
      }: Pick<SheetComponentsProps, 'sheetType'>) {
        const onMounted = jest.fn((instance) => {
          spreadSheet = instance;
        });

        return (
          <SheetComponent
            sheetType={sheetType}
            options={{ width: 200, height: 200 }}
            dataCfg={null as unknown as S2DataConfig}
            onMounted={onMounted}
          />
        );
      }

      renderComponent(<Component sheetType={'pivot'} />, container);

      await waitFor(() => {
        expect(spreadSheet!).toBeInstanceOf(PivotSheet);
      });

      renderComponent(<Component sheetType={'table'} />, container);

      await waitFor(() => {
        expect(spreadSheet!).toBeInstanceOf(TableSheet);
      });
    });
  });
});
