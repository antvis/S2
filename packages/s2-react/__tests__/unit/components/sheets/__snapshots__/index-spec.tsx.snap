// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SheetComponent/> Tests Render Tests should render editable sheet by snapshot 1`] = `
<DocumentFragment>
  <div
    class="antv-s2-wrapper"
  >
    <div
      class="antv-s2-container"
      style="position: relative;"
    >
      <canvas
        height="400"
        style="width: 200px; height: 200px; touch-action: none; display: block; user-select: none;"
        width="400"
      />
    </div>
    <div
      class="antv-s2-drag-copy-point"
      id="spreadsheet-drag-copy-point"
      style="display: none; left: -999px; top: -999px; width: 8px;"
    >
      <div
        class="antv-s2-drag-copy-mask"
        style="inset: 0px; width: 0px; height: 0px;"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`<SheetComponent/> Tests Render Tests should render gridAnalysis sheet by snapshot 1`] = `
<DocumentFragment>
  <div
    class="antv-s2-wrapper"
  >
    <div
      class="antv-s2-container"
      style="position: relative;"
    >
      <canvas
        height="400"
        style="width: 200px; height: 200px; touch-action: none; display: block; user-select: none;"
        width="400"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`<SheetComponent/> Tests Render Tests should render pivot sheet by snapshot 1`] = `
<DocumentFragment>
  <div
    class="antv-s2-wrapper"
  >
    <div
      class="antv-s2-container"
      style="position: relative;"
    >
      <canvas
        height="400"
        style="width: 200px; height: 200px; touch-action: none; display: block; user-select: none;"
        width="400"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`<SheetComponent/> Tests Render Tests should render strategy sheet by snapshot 1`] = `
<DocumentFragment>
  <div
    class="antv-s2-wrapper"
  >
    <div
      class="antv-s2-container"
      style="position: relative;"
    >
      <canvas
        height="400"
        style="width: 200px; height: 200px; touch-action: none; display: block; user-select: none;"
        width="400"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`<SheetComponent/> Tests Render Tests should render table sheet by snapshot 1`] = `
<DocumentFragment>
  <div
    class="antv-s2-wrapper"
  >
    <div
      class="antv-s2-container"
      style="position: relative;"
    >
      <canvas
        height="400"
        style="width: 200px; height: 200px; touch-action: none; display: block; user-select: none;"
        width="400"
      />
    </div>
  </div>
</DocumentFragment>
`;
